WasmSectionParser should use tryReserveInitialCapacity instead of tryReserveCapacity.

MenloDorian opened the pull request:
#### 78ca85d1db78101d75e8500282808226b695a1e7
<pre>
WasmSectionParser should use tryReserveInitialCapacity instead of tryReserveCapacity.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260913">https://bugs.webkit.org/show_bug.cgi?id=260913</a>
rdar://114175821

Reviewed by Justin Michaud.

We already enforced ordered section parsing, which also enforces that a section type can only
be parsed once (with CustomSection being the only exception).  For each of these sections, there
are many cases where capacity for their vector stores can be allocated with
Vector::tryReserveInitialCapacity instead of Vector::tryReserveCapacity.
Vector::tryReserveInitialCapacity will be more efficient.

Added some RELEASE_ASSERTs to enforce the invariant that these section types are only parsed once.
Added Vector::growCapacityBy and Vector::tryGrowCapacityBy to more concisely express the idea
that we&apos;re growing the capacity by some increment.

Also fixed an under-allocation of m_info-&gt;typeSignatures and m_info-&gt;rtts in SectionParser::parseRecursionGroup.
This code is still under development and is not yet deployed on any platforms.

* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
(JSC::Wasm::SectionParser::parseImport):
(JSC::Wasm::SectionParser::parseFunction):
(JSC::Wasm::SectionParser::parseGlobal):
(JSC::Wasm::SectionParser::parseExport):
(JSC::Wasm::SectionParser::parseElement):
(JSC::Wasm::SectionParser::parseFunctionType):
(JSC::Wasm::SectionParser::parseStructType):
(JSC::Wasm::SectionParser::parseRecursionGroup):
(JSC::Wasm::SectionParser::parseData):
(JSC::Wasm::SectionParser::parseException):
(JSC::Wasm::SectionParser::parseCustom):
* Source/WTF/wtf/Vector.h:
(WTF::Vector::growCapacityBy):
(WTF::Vector::tryGrowCapacityBy):
(WTF::Malloc&gt;::growCapacityBy):

Canonical link: <a href="https://commits.webkit.org/267469@main">https://commits.webkit.org/267469@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/518dd949f04f11bf5181a3ad3940c48bed5dea06

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16659 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17428 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18441 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15620 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20212 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17123 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16857 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14419 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19226 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14488 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15103 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14380 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15477 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15269 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19571 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15889 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15870 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18216 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15060 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/4267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19427 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19437 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2059 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15698 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4110 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1320,6 +1324,23 @@ inline bool Vector<T, inlineCapacity, OverflowHandler, minCapacity, Malloc>::res
+    unsigned increment32 = static_cast<unsigned>(increment);
+    unsigned capacity32 = static_cast<unsigned>(capacity());
+    unsigned newCapacity = increment32 + capacity32;
+    if (increment32 < increment || newCapacity < capacity32) {
```
Does this check actually work?

MenloDorian replied:
Yes it does work.

justinmichaud reviewed (approved):
r=me

Commit:
WasmSectionParser should use tryReserveInitialCapacity instead of tryReserveCapacity.
https://bugs.webkit.org/show_bug.cgi?id=260913
rdar://114175821

Reviewed by Justin Michaud.

We already enforced ordered section parsing, which also enforces that a section type can only
be parsed once (with CustomSection being the only exception).  For each of these sections, there
are many cases where capacity for their vector stores can be allocated with
Vector::tryReserveInitialCapacity instead of Vector::tryReserveCapacity.
Vector::tryReserveInitialCapacity will be more efficient.

Added some RELEASE_ASSERTs to enforce the invariant that these section types are only parsed once.
Added Vector::growCapacityBy and Vector::tryGrowCapacityBy to more concisely express the idea
that we're growing the capacity by some increment.

Also fixed an under-allocation of m_info->typeSignatures and m_info->rtts in SectionParser::parseRecursionGroup.
This code is still under development and is not yet deployed on any platforms.

* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
(JSC::Wasm::SectionParser::parseImport):
(JSC::Wasm::SectionParser::parseFunction):
(JSC::Wasm::SectionParser::parseGlobal):
(JSC::Wasm::SectionParser::parseExport):
(JSC::Wasm::SectionParser::parseElement):
(JSC::Wasm::SectionParser::parseFunctionType):
(JSC::Wasm::SectionParser::parseStructType):
(JSC::Wasm::SectionParser::parseRecursionGroup):
(JSC::Wasm::SectionParser::parseData):
(JSC::Wasm::SectionParser::parseException):
(JSC::Wasm::SectionParser::parseCustom):
* Source/WTF/wtf/Vector.h:
(WTF::Vector::growCapacityBy):
(WTF::Vector::tryGrowCapacityBy):
(WTF::Malloc>::growCapacityBy):

Canonical link: https://commits.webkit.org/267469@main
